config/v1/types_infrastructure.go

```diff
@@ -514,6 +525,26 @@ type GCPPlatformStatus struct {
 	Region string `json:"region"`
 }
 
+// BareMetalPlatformLoadBalancer defines the load balancer used by the cluster on BareMetal platform.
+type BareMetalPlatformLoadBalancer struct {
```

Comment:
You need a `+union` tag on the union struct itself as well as the discriminator tag which you already have